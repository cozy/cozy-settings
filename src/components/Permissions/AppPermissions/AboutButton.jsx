import React from 'react'

import { useClient, generateWebLink } from 'cozy-client'
import AppLinker from 'cozy-ui/transpiled/react/AppLinker'
import CircleButton from 'cozy-ui/transpiled/react/CircleButton'
import Icon from 'cozy-ui/transpiled/react/Icon'
import InfoIcon from 'cozy-ui/transpiled/react/Icons/Info'
import { useI18n } from 'cozy-ui/transpiled/react/providers/I18n'

export const AboutButton = ({ appData }) => {
  const { t } = useI18n()
  const client = useClient()
  const { subdomain: subDomainType } = client.getInstanceOptions()
  const appWebRef =
    appData &&
    generateWebLink({
      cozyUrl: client.getStackClient().uri,
      slug: 'store',
      subDomainType,
      hash: `discover/${appData.slug}`
    })

  return (
    <AppLinker app={appData} href={appWebRef}>
      {({ onClick, href }) => (
        <CircleButton
          label={t('Permissions.about')}
          href={href}
          onClick={onClick}
        >
          <Icon icon={InfoIcon} />
        </CircleButton>
      )}
    </AppLinker>
  )
}
