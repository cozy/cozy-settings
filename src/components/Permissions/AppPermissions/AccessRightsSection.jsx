import React from 'react'
import { NavigationListSection } from 'cozy-ui/transpiled/react/NavigationList'
import ListItem from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItem'
import Typography from 'cozy-ui/transpiled/react/Typography'
import Divider from 'cozy-ui/transpiled/react/MuiCozyTheme/Divider'
import ListItemIcon, {
  mediumSize,
  largeSize
} from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemIcon'
import Icon from 'cozy-ui/transpiled/react/Icon'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import RightIcon from 'cozy-ui/transpiled/react/Icons/Right'
import CozyLockIcon from 'cozy-ui/transpiled/react/Icons/CozyLock'
import CozyReleaseIcon from 'cozy-ui/transpiled/react/Icons/CozyRelease'
import {
  displayPermissions,
  getPermissionIconName
} from '../helpers/permissionsHelper'
import withAllLocales from '../../../lib/withAllLocales'
import useBreakpoints from 'cozy-ui/transpiled/react/hooks/useBreakpoints'
import { useNavigate } from 'react-router-dom'

const AccessRightsSection = ({
  sortedPermissionsByName,
  t,
  isRemoteDoctypes
}) => {
  const { isDesktop } = useBreakpoints()
  const navigate = useNavigate()

  const openModal = permissionType => {
    navigate(`./details/${permissionType}`)
  }

  if (!sortedPermissionsByName || sortedPermissionsByName.length < 1) {
    return
  }
  return (
    <>
      {!isDesktop && (
        <Divider
          style={{
            height: '0.75rem',
            backgroundColor: 'var(--defaultBackgroundColor)'
          }}
        />
      )}
      <NavigationListSection>
        <ListItem>
          <ListItemIcon>
            <Icon
              icon={isRemoteDoctypes ? CozyReleaseIcon : CozyLockIcon}
              size={largeSize}
            />
          </ListItemIcon>
          <ListItemText
            primary={
              <Typography type="body2" style={{ fontWeight: 'bold' }}>
                {isRemoteDoctypes
                  ? t('Permissions.exit_rights')
                  : t('Permissions.limited_right_access')}
              </Typography>
            }
            secondary={
              isRemoteDoctypes
                ? t('Permissions.exit_rights_secondary')
                : t('Permissions.limited_right_access_secondary')
            }
          />
        </ListItem>
        <Divider />
        {sortedPermissionsByName.map(({ name, title, verbs, type }, index) => {
          const iconName = getPermissionIconName(type)
          return (
            <div key={name}>
              <ListItem button onClick={() => openModal(type)}>
                <ListItemIcon>
                  <Icon
                    icon={
                      require(`cozy-ui/transpiled/react/Icons/${iconName}`)
                        .default
                    }
                    size={mediumSize}
                  />
                </ListItemIcon>
                <ListItemText
                  primary={title}
                  secondary={t(displayPermissions(verbs))}
                />
                <Icon icon={RightIcon} />
              </ListItem>
              {index !== sortedPermissionsByName.length - 1 && (
                <Divider variant="inset" />
              )}
            </div>
          )
        })}
      </NavigationListSection>
    </>
  )
}

export default withAllLocales(AccessRightsSection)
