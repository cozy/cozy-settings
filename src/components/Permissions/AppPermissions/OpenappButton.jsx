import React from 'react'
import CircleButton from 'cozy-ui/transpiled/react/CircleButton'
import Icon from 'cozy-ui/transpiled/react/Icon'
import OpenappIcon from 'cozy-ui/transpiled/react/Icons/Openapp'
import AppLinker from 'cozy-ui/transpiled/react/AppLinker'
import { useClient, generateWebLink } from 'cozy-client'
import { useI18n } from 'cozy-ui/transpiled/react/providers/I18n'

export const OpenappButton = ({ type, appData }) => {
  const { t } = useI18n()
  const client = useClient()
  const { subdomain: subDomainType } = client.getInstanceOptions()

  let appWebRef = ''
  if (type === 'konnector') {
    appWebRef =
      appData &&
      generateWebLink({
        cozyUrl: client.getStackClient().uri,
        slug: 'home',
        subDomainType,
        hash: `connected/${appData.slug}`
      })
  } else {
    appWebRef = appData.links?.related
  }

  return (
    <AppLinker app={appData} href={appWebRef}>
      {({ onClick, href }) => (
        <CircleButton
          label={t('Permissions.open')}
          href={href}
          onClick={onClick}
        >
          <Icon icon={OpenappIcon} />
        </CircleButton>
      )}
    </AppLinker>
  )
}
