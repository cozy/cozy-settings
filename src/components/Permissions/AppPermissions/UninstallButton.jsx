import React from 'react'
import CircleButton from 'cozy-ui/transpiled/react/CircleButton'
import Icon from 'cozy-ui/transpiled/react/Icon'
import TrashIcon from 'cozy-ui/transpiled/react/Icons/Trash'
import AppLinker from 'cozy-ui/transpiled/react/AppLinker'
import { useI18n } from 'cozy-ui/transpiled/react/providers/I18n'
import { useClient, generateWebLink } from 'cozy-client'

export const UninstallButton = ({ appData }) => {
  const { t } = useI18n()
  const client = useClient()
  const { subdomain: subDomainType } = client.getInstanceOptions()
  const appWebRef =
    appData &&
    generateWebLink({
      cozyUrl: client.getStackClient().uri,
      slug: 'store',
      subDomainType,
      hash: `discover/${appData.slug}/uninstall`
    })

  return (
    <AppLinker app={appData} href={appWebRef}>
      {({ onClick, href }) => (
        <CircleButton
          label={t('Permissions.uninstall')}
          href={href}
          onClick={onClick}
        >
          <Icon icon={TrashIcon} />
        </CircleButton>
      )}
    </AppLinker>
  )
}
