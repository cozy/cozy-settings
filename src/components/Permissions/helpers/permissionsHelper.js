import PERMISSIONS_ICONS from '../../../config/permissionsIcons.json'

export const displayPermissions = verbs => {
  return !verbs || (verbs.length > 1 && verbs.includes('GET'))
    ? 'Permissions.readAndWrite'
    : verbs.length === 1 && verbs.includes('GET')
    ? 'Permissions.read'
    : 'Permissions.write'
}

export const getPermissionIconName = type => {
  return PERMISSIONS_ICONS[type] || 'Fallback'
}

export const completePermission = (apps, konnectors) => {
  const hiddenSlugs = ['home', 'store', 'settings']
  let slugs = []
  const permissionsToDisplay = {}
  if (apps?.data?.length > 0 || konnectors?.data?.length > 0) {
    slugs = [
      ...apps.data.filter(({ slug }) => !hiddenSlugs.includes(slug)),
      ...konnectors.data
    ]

    for (const { slug, permissions } of slugs) {
      Object.values(permissions).map(({ type }) => {
        if (!permissionsToDisplay[type]) {
          permissionsToDisplay[type] = []
        }
        permissionsToDisplay[type].push(slug)
      })
    }
  }
  return permissionsToDisplay
}

export const getPermissionsVerbsByType = (permissions, type) => {
  let verbs
  const permissionsArray = Object.entries(permissions)
  for (let i = 0; i < permissionsArray.length; i++) {
    if (permissionsArray[i][1].type === type) {
      verbs = permissionsArray[i][1].verbs
    }
  }
  return verbs
}

export const isNotLastItem = (itemId, list) => {
  if (list[0].name) {
    return itemId !== list[list.length - 1].name
  } else {
    return itemId !== list[list.length - 1].id
  }
}

export const filterPermissions = (remoteDoctypes, matchingQueryResult) => {
  let limitedRightAccess = []
  let exitRights = []
  if (remoteDoctypes && matchingQueryResult?.data?.permissions) {
    const permissions = Object.entries(matchingQueryResult.data.permissions)
    for (let i = 0; i < permissions.length; i++) {
      if (remoteDoctypes.includes(permissions[i][1].type)) {
        exitRights.push(permissions[i])
      } else {
        limitedRightAccess.push(permissions[i])
      }
    }
  }
  return { limitedRightAccess, exitRights }
}

export const sortPermissionsByName = (completePermission, permissions, t) => {
  return permissions
    .map(([name, value]) => {
      const type = value.type
      const perm = t('CozyPermissions.Permissions.' + type)
      return completePermission(name, perm, value, type)
    })
    .sort((a, b) => {
      return a.title.localeCompare(b.title)
    })
}

export const completeAppPermission = (
  name,
  permission,
  { description, verbs },
  type
) => {
  const completedPermission = {
    name,
    type,
    title: permission
  }
  if (description) {
    completedPermission.description = description
  }
  if (verbs) {
    completedPermission.verbs = verbs
  }
  return completedPermission
}

export const sortDataByDate = queryResult => {
  return queryResult.data.sort(
    (a, b) => new Date(b.created_at) - new Date(a.created_at)
  )
}
