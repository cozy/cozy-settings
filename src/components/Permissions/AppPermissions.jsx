import React from 'react'

import { APPS_DOCTYPE, KONNECTORS_DOCTYPE } from 'doctypes'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import Typography from 'cozy-ui/transpiled/react/Typography'
import Page from 'components/Page'
import PageTitle from 'components/PageTitle'
import { Link, useParams } from 'react-router-dom'
import {
  displayPermissions,
  getPermissionIconName
} from '../../containers/helpers/permissionsHelper'
import Icon from 'cozy-ui/transpiled/react/Icon'
import IconButton from 'cozy-ui/transpiled/react/IconButton'
import PreviousIcon from 'cozy-ui/transpiled/react/Icons/Previous'
import ListItemIcon, {
  smallSize,
  mediumSize
} from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemIcon'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import ListItem from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItem'
import Divider from 'cozy-ui/transpiled/react/MuiCozyTheme/Divider'
import { routes } from 'constants/routes'
import { OpenappButton } from './OpenappButton'
import withAllLocales from '../../lib/withAllLocales'
import ListItemSecondaryAction from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemSecondaryAction'
import RightIcon from 'cozy-ui/transpiled/react/Icons/Right'
import NavigationList from 'cozy-ui/transpiled/react/NavigationList'
import CozyClient, {
  Q,
  useQuery,
  isQueryLoading,
  hasQueryBeenLoaded
} from 'cozy-client'

export const completePermission = (
  name,
  permission,
  { description, verbs },
  type
) => {
  const completedPermission = {
    name,
    type,
    title: permission
  }
  if (description) {
    completedPermission.description = description
  }
  if (verbs) {
    completedPermission.verbs = verbs
  }
  return completedPermission
}

const PermissionsApplication = ({ t }) => {
  const { slug: slugName } = useParams()
  const THIRTY_SECONDS = 30 * 1000

  const queryResultApps = useQuery(
    Q(APPS_DOCTYPE).getById('io.cozy.apps/' + slugName),
    {
      as: 'io.cozy.apps/' + slugName,
      fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
    }
  )

  const queryResultKonnectors = useQuery(
    Q(KONNECTORS_DOCTYPE).getById('io.cozy.konnectors/' + slugName),
    {
      as: 'io.cozy.konnectors/' + slugName,
      fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
    }
  )

  let matchingQueryResult
  if (queryResultApps?.data?.length > 0) {
    matchingQueryResult = queryResultApps
  }
  if (queryResultKonnectors?.data?.length > 0) {
    matchingQueryResult = queryResultKonnectors
  }
  const sortPermissionsByName = queryResult => {
    return Object.entries(queryResult.data[0].attributes.permissions)
      .map(([name, value]) => {
        const type = value.type
        const perm = t('CozyPermissions.Permissions.' + type)
        return completePermission(name, perm, value, type)
      })
      .sort((a, b) => {
        return a.title.localeCompare(b.title)
      })
  }

  const isKonnector = type => type === 'io.cozy.konnectors'

  let sortedPermissionsByName =
    matchingQueryResult && sortPermissionsByName(matchingQueryResult)

  const isNotLastItem = name => {
    return (
      name !== sortedPermissionsByName[sortedPermissionsByName.length - 1].name
    )
  }

  return (
    <Page narrow>
      {(isQueryLoading(queryResultApps) &&
        !hasQueryBeenLoaded(queryResultApps)) ||
      (isQueryLoading(queryResultKonnectors) &&
        !hasQueryBeenLoaded(queryResultKonnectors)) ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : queryResultApps.fetchStatus === 'failed' &&
        queryResultKonnectors.fetchStatus === 'failed' ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <div>
          <IconButton className="u-mr-half" href="#/permissions">
            <Icon icon={PreviousIcon} size={16} />
          </IconButton>
          <NavigationList>
            <PageTitle>{slugName.toUpperCase()}</PageTitle>
            <OpenappButton
              type={
                isKonnector(matchingQueryResult.data[0].type)
                  ? 'konnector'
                  : 'app'
              }
              matchingQueryResultData={matchingQueryResult.data[0]}
            />

            {sortedPermissionsByName.map(({ name, title, verbs, type }) => {
              const iconName = getPermissionIconName(type)
              return (
                <Link
                  to={`${routes.appList}/${slugName}/${name}`}
                  key={name}
                  style={{ textDecoration: 'none', color: 'black' }}
                >
                  <ListItem button>
                    <ListItemIcon>
                      <Icon
                        icon={
                          require(`cozy-ui/transpiled/react/Icons/${iconName}`)
                            .default
                        }
                        size={mediumSize}
                      />
                    </ListItemIcon>
                    <ListItemText
                      primary={title}
                      secondary={t(displayPermissions(verbs))}
                    />
                    <ListItemSecondaryAction style={{ color: 'grey' }}>
                      <Icon
                        icon={RightIcon}
                        size={smallSize}
                        className="u-mr-1"
                      />
                    </ListItemSecondaryAction>
                  </ListItem>
                  {isNotLastItem(name) && <Divider variant="inset" />}
                </Link>
              )
            })}
          </NavigationList>
        </div>
      )}
    </Page>
  )
}

export default withAllLocales(PermissionsApplication)
