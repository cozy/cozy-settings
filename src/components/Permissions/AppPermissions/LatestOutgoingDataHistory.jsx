import React, { useState } from 'react'

import { Dialog as DialogComponent } from 'cozy-ui/transpiled/react/CozyDialogs'
import Divider from 'cozy-ui/transpiled/react/Divider'
import Icon from 'cozy-ui/transpiled/react/Icon'
import RightIcon from 'cozy-ui/transpiled/react/Icons/Right'
import RiseIcon from 'cozy-ui/transpiled/react/Icons/Rise'
import ListItem from 'cozy-ui/transpiled/react/ListItem'
import ListItemIcon, {
  smallSize,
  mediumSize
} from 'cozy-ui/transpiled/react/ListItemIcon'
import ListItemSecondaryAction from 'cozy-ui/transpiled/react/ListItemSecondaryAction'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import { NavigationListSection } from 'cozy-ui/transpiled/react/NavigationList'
import Typography from 'cozy-ui/transpiled/react/Typography'
import useBreakpoints from 'cozy-ui/transpiled/react/providers/Breakpoints'
import { useI18n } from 'cozy-ui/transpiled/react/providers/I18n'

import withAllLocales from '../../../lib/withAllLocales'
import { sortDataByDate } from '../helpers/permissionsHelper'

const LatestOutgoingDataHistory = ({
  queryResultRemoteRequests,
  slugName,
  t
}) => {
  const [modalOpened, setModalOpened] = useState(false)
  const [modalData, setModalData] = useState()
  const { f } = useI18n()
  const { isDesktop } = useBreakpoints()

  const openModal = data => {
    setModalOpened(true)
    let formatedData = { ...data }
    delete formatedData._id
    delete formatedData._rev
    setModalData(formatedData)
  }

  const formatDate = ({ f, date }) => {
    return f(date, 'dd LLLL yyyy')
  }

  const handleClose = () => setModalOpened(false)

  return (
    <>
      {!isDesktop && (
        <Divider
          style={{
            height: '0.75rem',
            backgroundColor: 'var(--defaultBackgroundColor)'
          }}
        />
      )}
      <NavigationListSection>
        <ListItem ellipsis={false}>
          <Typography variant="h5">
            {t('Permissions.latest_data_releases')}
          </Typography>
        </ListItem>
        <Divider />
        {sortDataByDate(queryResultRemoteRequests).map((data, index) => {
          return (
            <div key={data.id}>
              <ListItem button onClick={() => openModal(data)} ellipsis={false}>
                <ListItemIcon>
                  <Icon
                    icon={RiseIcon}
                    size={mediumSize}
                    color="var(--warningColor)"
                    rotate={90}
                  />
                </ListItemIcon>
                <ListItemText
                  primary={t(`CozyPermissions.${data.doctype}`)}
                  secondary={formatDate({
                    f,
                    date: new Date(data.created_at)
                  })}
                />
                <ListItemSecondaryAction>
                  <Icon
                    icon={RightIcon}
                    size={smallSize}
                    className="u-mr-1"
                    style={{ color: 'var(--secondaryTextColor)' }}
                  />
                </ListItemSecondaryAction>
              </ListItem>
              {index !==
                sortDataByDate(queryResultRemoteRequests).length - 1 && (
                <Divider variant="inset" />
              )}
            </div>
          )
        })}
      </NavigationListSection>
      <DialogComponent
        open={modalOpened}
        onClose={handleClose}
        title={t('Permissions.monthly_statistics')}
        content={
          <>
            <p>
              {modalData &&
                t('Permissions.monthly_stats_phrase', {
                  app: slugName.toUpperCase(),
                  date: formatDate({
                    f,
                    date: new Date(modalData.created_at)
                  })
                })}
            </p>
            <code>
              <pre>{modalData && JSON.stringify(modalData, null, 2)}</pre>
            </code>
          </>
        }
      />
    </>
  )
}

export default withAllLocales(LatestOutgoingDataHistory)
