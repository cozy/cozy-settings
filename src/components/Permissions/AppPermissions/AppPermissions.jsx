import React, { useMemo } from 'react'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import Typography from 'cozy-ui/transpiled/react/Typography'
import Page from 'components/Page'
import PageTitle from 'components/PageTitle'
import { useParams } from 'react-router-dom'
import AppIcon from 'cozy-ui/transpiled/react/AppIcon'
import Icon from 'cozy-ui/transpiled/react/Icon'
import IconButton from 'cozy-ui/transpiled/react/IconButton'
import PreviousIcon from 'cozy-ui/transpiled/react/Icons/Previous'
import { OpenappButton } from 'components/Permissions/OpenappButton'
import { AboutButton } from 'components/Permissions/AboutButton'
import withAllLocales from 'lib/withAllLocales'
import NavigationList from 'cozy-ui/transpiled/react/NavigationList'
import { isQueryLoading, hasQueryBeenLoaded } from 'cozy-client'
import AccessRightsSection from 'components/Permissions/AccessRightsSection'
import LatestOutgoingDataHistory from 'components/Permissions/LatestOutgoingDataHistory'
import {
  filterPermissions,
  sortPermissionsByName,
  completeAppPermission,
  filterRemoteRequests
} from 'components/Permissions/helpers/permissionsHelper'
import useFetchJSON from 'cozy-client/dist/hooks/useFetchJSON'
import {
  buildAppsQueryBySlug,
  buildKonnectorsQueryBySlug,
  buildRemoteRequestsQuery
} from 'lib/queries'
import { useQuery } from 'cozy-client'

const AppPermissions = ({ t }) => {
  const { slug: slugName } = useParams()

  const appsQuery = buildAppsQueryBySlug(slugName)
  const queryResultApps = useQuery(appsQuery.definition, appsQuery.options)
  const konnectorsQuery = buildKonnectorsQueryBySlug(slugName)
  const queryResultKonnectors = useQuery(
    konnectorsQuery.definition,
    konnectorsQuery.options
  )
  const remoteQuery = buildRemoteRequestsQuery()
  const queryResultRemoteRequests = useQuery(
    remoteQuery.definition,
    remoteQuery.options
  )

  const { data: remoteDoctypes, ...queryResultRemoteDoctypes } = useFetchJSON(
    'GET',
    '/remote/_all_doctypes'
  )

  let matchingQueryResult
  if (queryResultApps.data) {
    matchingQueryResult = queryResultApps
  }
  if (queryResultKonnectors.data) {
    matchingQueryResult = queryResultKonnectors
  }

  const filteredPermissions = useMemo(
    () => filterPermissions(remoteDoctypes, matchingQueryResult),
    [remoteDoctypes, matchingQueryResult]
  )

  const limitedRightAccess = filteredPermissions.limitedRightAccess
  const exitRights = filteredPermissions.exitRights

  const sortedLimitedRightAccessPermissionsByName = sortPermissionsByName(
    completeAppPermission,
    limitedRightAccess,
    t
  )
  const sortedExitPermissionsByName = sortPermissionsByName(
    completeAppPermission,
    exitRights,
    t
  )

  const isKonnector = type => type === 'io.cozy.konnectors'

  const filteredRemoteRequests = filterRemoteRequests(
    queryResultRemoteRequests,
    slugName
  )

  return (
    <Page narrow>
      {(isQueryLoading(queryResultApps) &&
        !hasQueryBeenLoaded(queryResultApps)) ||
      (isQueryLoading(queryResultKonnectors) &&
        !hasQueryBeenLoaded(queryResultKonnectors)) ||
      (isQueryLoading(queryResultRemoteRequests) &&
        !hasQueryBeenLoaded(queryResultRemoteRequests)) ||
      (isQueryLoading(queryResultRemoteDoctypes) &&
        !hasQueryBeenLoaded(queryResultRemoteDoctypes)) ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : (queryResultApps.fetchStatus === 'failed' &&
          queryResultKonnectors.fetchStatus === 'failed') ||
        queryResultRemoteDoctypes.fetchStatus === 'error' ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <div>
          <IconButton className="u-mr-half" href="#/permissions">
            <Icon icon={PreviousIcon} size={16} />
          </IconButton>
          <div className="u-flex u-flex-column u-flex-items-center">
            <div style={{ width: '48px' }}>
              <AppIcon app={slugName} />
            </div>
            <PageTitle>{slugName.toUpperCase()}</PageTitle>
            <div>
              <OpenappButton
                type={
                  isKonnector(matchingQueryResult.data.type)
                    ? 'konnector'
                    : 'app'
                }
                matchingQueryResultData={matchingQueryResult.data}
              />
              <AboutButton matchingQueryResultData={matchingQueryResult.data} />
            </div>
          </div>
          <NavigationList>
            {sortedLimitedRightAccessPermissionsByName.length > 0 && (
              <AccessRightsSection
                sortedPermissionsByName={
                  sortedLimitedRightAccessPermissionsByName
                }
                slugName={slugName}
                t={t}
                isRemoteDoctypes={false}
              />
            )}

            {sortedExitPermissionsByName.length > 0 && (
              <AccessRightsSection
                sortedPermissionsByName={sortedExitPermissionsByName}
                slugName={slugName}
                t={t}
                isRemoteDoctypes={true}
              />
            )}

            {filteredRemoteRequests?.length > 0 && (
              <LatestOutgoingDataHistory
                queryResultRemoteRequests={filteredRemoteRequests}
                slugName={slugName}
                t={t}
              />
            )}
          </NavigationList>
        </div>
      )}
    </Page>
  )
}

export default withAllLocales(AppPermissions)
