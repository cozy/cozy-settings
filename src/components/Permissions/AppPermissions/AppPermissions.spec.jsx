import React from 'react'
import { render } from '@testing-library/react'

import {
  Q,
  useQuery,
  useClient,
  isQueryLoading,
  hasQueryBeenLoaded
} from 'cozy-client'
import useFetchJSON from 'cozy-client/dist/hooks/useFetchJSON'
import { BreakpointsProvider } from 'cozy-ui/transpiled/react/hooks/useBreakpoints'

import AppPermissions from 'components/Permissions/AppPermissions/AppPermissions'
import useAppsOrKonnectorsBySlug from 'components/Permissions/hooks/useAppsOrKonnectorsBySlug'
import { completeAppPermission } from 'components/Permissions/helpers/permissionsHelper'

jest.mock('cozy-ui/transpiled/react/I18n/withLocales', () => {
  return () => Component => {
    const t = text => text
    const match = { params: { slug: 'Drive' } }
    // eslint-disable-next-line react/display-name
    return () => <Component match={match} t={t} />
  }
})

jest.mock('cozy-ui/transpiled/react/NavigationList', () => {
  return {
    __esModule: true,
    default: ({ children }) => (
      <div data-testid="NavigationList">{children}</div>
    ),
    NavigationListSection: ({ children }) => (
      <div data-testid="NavigationListSection">{children}</div>
    )
  }
})

jest.mock('react-router-dom', () => {
  return {
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({ slug: 'Drive' }),
    Link: ({ narrow, children }) => (
      <div data-testid="page" data-narrow={narrow}>
        {children}
      </div>
    )
  }
})

jest.mock('cozy-ui/transpiled/react/CozyDialogs', () => {
  return {
    Dialog: () => (
      <div
        data-testid="Dialog"
        data-open="open"
        data-onClose="onClose"
        data-title="title"
        data-content="content"
      />
    )
  }
})

jest.mock('cozy-client', () => ({
  ...jest.requireActual('cozy-client'),
  hasQueryBeenLoaded: jest.fn(),
  Q: jest.fn(),
  useQuery: jest.fn(),
  isQueryLoading: jest.fn(),
  useClient: jest.fn()
}))

jest.mock('components/Page', () => {
  return ({ narrow, children }) => (
    <div data-testid="page" data-narrow={narrow}>
      {children}
    </div>
  )
})

jest.mock('components/PageTitle', () => {
  return ({ children }) => <div data-testid="pageTitle">{children}</div>
})
jest.mock('cozy-ui/transpiled/react/Spinner', () => {
  return ({ size }) => <div data-testid="Spinner" data-size={size}></div>
})
jest.mock('cozy-ui/transpiled/react/CircleButton', () => {
  return () => <div data-testid="CircleButton"></div>
})
jest.mock('cozy-ui/transpiled/react/I18n', () => {
  return { useI18n: () => ({ t: x => x }) }
})
jest.mock('cozy-client/dist/hooks/useFetchJSON', () => ({
  __esModule: true,
  default: jest.fn()
}))

jest.mock('../hooks/useAppsOrKonnectorsBySlug')

describe('AppPermissions', () => {
  beforeEach(() => {
    const queryResult = {
      fetchStatus: 'loaded',
      data: {
        permissions: {
          files: {
            type: 'io.cozy.files',
            description: 'Required to access the files'
          },
          allFiles: {
            type: 'io.cozy.files.*',
            description: 'Required to access the files'
          },
          apps: {
            type: 'io.cozy.apps',
            description:
              'Required by the cozy-bar to display the icons of the apps',
            verbs: ['GET']
          },
          accounts: {
            type: 'io.cozy.accounts',
            description: 'Required to access accounts',
            verbs: ['POST']
          }
        }
      }
    }
    isQueryLoading.mockReturnValue(true)
    hasQueryBeenLoaded.mockReturnValue(true)
    Q.mockReturnValue({ getById: () => 'kfrf' })
    useQuery.mockReturnValue(queryResult)
    useClient.mockReturnValue({
      getStackClient: () => {
        return { uri: 'http://cozy.localhost:8080' }
      },
      getInstanceOptions: () => {
        return {
          subdomain: 'flat'
        }
      }
    })
  })
  it('should display slugName when query has been loaded', () => {
    hasQueryBeenLoaded.mockReturnValue(true)
    useAppsOrKonnectorsBySlug.mockReturnValue({
      isResultLoading: false,
      hasQueryFailed: false,
      result: { fetchStatus: 'loaded', data: {} }
    })
    useFetchJSON.mockReturnValue({ data: ['doctype1', 'doctype2'] })
    const { container } = render(
      <BreakpointsProvider>
        <AppPermissions />
      </BreakpointsProvider>
    )
    expect(container).toMatchSnapshot()
  })

  it('should render a spinner when query is loading and has not been loaded', () => {
    hasQueryBeenLoaded.mockReturnValue(false)
    useAppsOrKonnectorsBySlug.mockReturnValue({
      isResultLoading: true,
      hasQueryFailed: false,
      result: { fetchStatus: 'loading', data: {} }
    })
    const { queryByTestId } = render(
      <BreakpointsProvider>
        <AppPermissions />
      </BreakpointsProvider>
    )
    expect(queryByTestId('Spinner')).toBeTruthy()
  })

  it('should display slugName when query is not loading', () => {
    useAppsOrKonnectorsBySlug.mockReturnValue({
      isResultLoading: false,
      hasQueryFailed: false,
      result: { fetchStatus: 'loaded', data: {} }
    })
    const { queryByText } = render(
      <BreakpointsProvider>
        <AppPermissions />
      </BreakpointsProvider>
    )
    expect(queryByText('DRIVE')).toBeInTheDocument()
  })

  it('should render Permissions.failedRequest when query status is failed', () => {
    useAppsOrKonnectorsBySlug.mockReturnValue({
      isResultLoading: false,
      hasQueryFailed: true,
      result: { fetchStatus: 'loaded', data: {} }
    })
    const { queryByText } = render(
      <BreakpointsProvider>
        <AppPermissions />
      </BreakpointsProvider>
    )
    expect(queryByText('Permissions.failedRequest')).toBeTruthy()
  })

  describe('completeAppPermission', () => {
    it('should add description in permissions', () => {
      const key = 'contactsAccounts'
      const permission = 'Comptes associés aux contacts'
      const value = {
        description: 'blabla',
        verbs: ['GET', 'POST']
      }
      const resultat = completeAppPermission(key, permission, value)
      expect(resultat).toEqual({
        description: 'blabla',
        name: 'contactsAccounts',
        title: 'Comptes associés aux contacts',
        verbs: ['GET', 'POST']
      })
    })
  })
})
