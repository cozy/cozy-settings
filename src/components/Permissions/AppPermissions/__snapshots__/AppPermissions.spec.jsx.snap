// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppPermissions should display slugName when query has been loaded 1`] = `
<div>
  <div
    data-testid="page"
  >
    <div>
      <a
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root u-mr-half large"
        href="#/permissions"
        tabindex="0"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            class="styles__icon___23x3R"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M3.414 7h11.58C15.548 7 16 7.444 16 8c0 .552-.45 1-1.007 1H3.414l5.293 5.293a1 1 0 01-1.414 1.414l-7-7a1 1 0 010-1.414l7-7a1 1 0 111.414 1.414L3.414 7z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </a>
      <div
        class="u-flex u-flex-column u-flex-items-center"
      >
        <div
          style="width: 48px;"
        >
          <svg
            class="styles__c-app-icon___2_O40 styles__c-app-icon-default___3CEmt styles__icon___23x3R"
            height="100%"
            style="fill: var(--coolGrey);"
            viewBox="0 0 16 16"
            width="100%"
          >
            <path
              d="M1 11.009V5.5c0-.55.39-.773.872-.498l5.256 3.003c.476.272.872.944.872 1.495v5.508c0 .549-.39.772-.872.497l-5.256-3.003C1.396 12.231 1 11.56 1 11.01zm15 0c0 .55-.396 1.222-.872 1.494l-5.256 3.003c-.481.275-.872.052-.872-.497V9.5c0-.55.396-1.223.872-1.495l5.256-3.003c.481-.275.872-.052.872.498v5.508zM9.35 6.982c-.47.288-1.237.284-1.7 0l-4.8-2.954c-.47-.29-.463-.732.027-.995L7.623.477c.485-.261 1.264-.264 1.754 0l4.746 2.556c.485.26.49.71.027.995l-4.8 2.954z"
            />
          </svg>
        </div>
        <div
          data-testid="pageTitle"
        >
          DRIVE
        </div>
        <div>
          <div
            data-testid="CircleButton"
          />
          <div
            data-testid="CircleButton"
          />
          <div
            data-testid="CircleButton"
          />
        </div>
      </div>
      <div
        data-testid="NavigationList"
      />
    </div>
  </div>
</div>
`;
